src/main/scala/com/microsoft/hyperspace/util/DataAnalysisUtils.scala

```diff
@@ -0,0 +1,278 @@
+
+    s"""
+    <!DOCTYPE html>
+    <meta charset="utf-8">
+    <script src="https://d3js.org/d3.v4.js"></script>
+    <div id="my_dataviz"></div>
+
+    <script>
+    var margin = {"top": 20, "right": 30, "bottom": 25, "left": 50 }
+    var width = 500;
+    var height = 500;
+    var rectWidth = ${500 / values.length};
+
+    var data = [$valueStr];
+
+    var xMax = ${values.length} * rectWidth;
+    var xScale = d3.scaleLinear()
+      .domain([0, xMax])
+      .range([margin.left, width - margin.right]);
+
+    var yMax = $yMax;
+    var yScale = d3.scaleLinear()
+      .domain([0, yMax])
+      .range([height - margin.bottom, margin.top]);
+
+    var svg = d3.select('svg');
+
+    var rect = svg.selectAll('rect')
+      .data(data)
+      .enter().append('rect')
+      .attr('x', function(d, i) {
+        return xScale(i * rectWidth)})
+      .attr('y', function(d) {
+        return yScale(d)})
+      .attr('width', xScale(rectWidth) - margin.left)
+      .attr('height', function(d) {
+        return height - margin.bottom - yScale(d)})
+      .attr('fill', "#AED6F1")
+      .attr('margin', 0);
+
+    var xAxis = d3.axisBottom()
+      .scale(xScale)
+      .tickFormat((d, i) => ['Min value', 'Max value'][i])
+      .ticks(1);
+    var yAxis = d3.axisLeft()
+      .scale(yScale)
+      .tickFormat(d3.format('d'))
+      .ticks($yTickNum);
+
+    svg.append('g')
+      .attr('transform', 'translate(' + [0, height - margin.bottom] + ')')
+      .call(xAxis);
+    svg.append('g')
+      .attr('transform', 'translate(' + [margin.left, 0] + ')')
+      .call(yAxis);
+
+    svg.append("text")
+      .attr("text-anchor", "middle")
+      .attr("transform", "translate(" + 15 + "," + (height / 2) + ")rotate(-90)")
+      .style("font-size", "15px")
+      .text("Max. number of files to lookup");
+
+    svg.append("text")
+      .attr("text-anchor", "middle")
+      .attr("transform", "translate(" + (width / 2) + "," + (height - 4) + ")")
+      .style("font-size", "15px")
+      .text("Column value range");
+
+    </script>
+    <body><svg width=500 height=500></svg></body>
+    """
+  }
+
+  def minMaxAnalysis(df: DataFrame, colName: String, htmlFormat: Boolean = false): String = {
+    val dataType = df.schema.fields.find(_.name.equals(colName)).get.dataType
+    val numeric = try {
+      TypeUtils.getNumeric(dataType)
+    } catch {
+      case _: Exception =>
+        throw HyperspaceException(
+          s"minMaxAnalysis only supports Numeric type. colName=$colName, dataType=$dataType")
+    }
+
+    val distFigureLen = if (htmlFormat) 20 else 50
+    val newline = if (htmlFormat) "</br>" else System.lineSeparator()
+    val res = df.inputFiles.par.map { f =>
+      val file = df.sparkSession.read.parquet(f)
+      f -> file.agg(min(col(colName)), max(col(colName))).collect.toSeq.headOption
+    }.toIndexedSeq.filter(_._2.isDefined)
+
+    /* groupBy shuffles all data, though it can be done with one spark job.
```

Comment:
Are you sure about this? There is no reason that Spark should shuffle "all data" when data is grouped by `input_file_name()`. All that shuffled is just the aggregated data, which is much less than the source data.

Reply:
yea seems spark does the aggregation before groupBy exchange. I'll remove the comment.